# Original Code

In [1]:
"""
This Python script builds and trains a Recurrent Neural Network (RNN) to classify the sentiment of
movie reviews using the IMDb dataset. The model uses an Embedding layer and a SimpleRNN layer. 
It is compiled and trained using the TensorFlow and Keras libraries.
"""
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, Dense
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing import sequence

# Cargar y preprocesar los datos de IMDB
max_features = 10000
maxlen = 100

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=max_features)
X_train = sequence.pad_sequences(X_train, maxlen=maxlen)
X_test = sequence.pad_sequences(X_test, maxlen=maxlen)

# Crear y entrenar el modelo RNN
model = Sequential()
model.add(Embedding(max_features, 32, input_length=maxlen))
model.add(SimpleRNN(32, activation='tanh'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.2)

# Evaluar el modelo
accuracy = model.evaluate(X_test, y_test)
print(f'Test accuracy: {accuracy[1]}')

17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Epoch 1/5


C:\Users\luism\anaconda\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


313/313 ━━━━━━━━━━━━━━━━━━━━ 6s 12ms/step - accuracy: 0.6264 - loss: 0.6150 - val_accuracy: 0.8068 - val_loss: 0.4366
Epoch 2/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 5s 16ms/step - accuracy: 0.8803 - loss: 0.3103 - val_accuracy: 0.8336 - val_loss: 0.4205
Epoch 3/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 5s 15ms/step - accuracy: 0.9323 - loss: 0.1909 - val_accuracy: 0.8144 - val_loss: 0.4516
Epoch 4/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 5s 15ms/step - accuracy: 0.9701 - loss: 0.0948 - val_accuracy: 0.8170 - val_loss: 0.4812
Epoch 5/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 5s 15ms/step - accuracy: 0.9900 - loss: 0.0409 - val_accuracy: 0.8012 - val_loss: 0.5575
782/782 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7945 - loss: 0.5793
Test accuracy: 0.7950000166893005


# Modified Code

In [2]:
from tensorflow.keras.layers import LSTM, Dropout

# Load and preprocess the IMDb data
max_features = 10000  # Vocabulary size
maxlen = 100  # Cut sequences after this many words

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=max_features)
X_train = sequence.pad_sequences(X_train, maxlen=maxlen)
X_test = sequence.pad_sequences(X_test, maxlen=maxlen)

# Create and train the modified RNN model
model = Sequential()
model.add(Embedding(max_features, 64, input_length=maxlen))  # Increased embedding size
model.add(Dropout(0.5))  # Dropout after embedding layer
model.add(LSTM(32, activation='tanh'))  # Replaced SimpleRNN with LSTM
model.add(Dropout(0.5))  # Dropout after LSTM layer
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
history = model.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.2)

# Evaluate the model
accuracy = model.evaluate(X_test, y_test)
print(f'Test accuracy: {accuracy[1]}')

Epoch 1/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 11s 27ms/step - accuracy: 0.6519 - loss: 0.5919 - val_accuracy: 0.8476 - val_loss: 0.3513
Epoch 2/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 10s 25ms/step - accuracy: 0.8854 - loss: 0.2874 - val_accuracy: 0.8540 - val_loss: 0.3361
Epoch 3/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 8s 24ms/step - accuracy: 0.9152 - loss: 0.2301 - val_accuracy: 0.8438 - val_loss: 0.3663
Epoch 4/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 8s 26ms/step - accuracy: 0.9307 - loss: 0.1930 - val_accuracy: 0.8424 - val_loss: 0.4205
Epoch 5/5
313/313 ━━━━━━━━━━━━━━━━━━━━ 11s 33ms/step - accuracy: 0.9455 - loss: 0.1530 - val_accuracy: 0.8332 - val_loss: 0.5057
782/782 ━━━━━━━━━━━━━━━━━━━━ 7s 9ms/step - accuracy: 0.8271 - loss: 0.5233
Test accuracy: 0.8277999758720398


**Explanation of Changes:**

* Embedding size was increased from 32 to 64 to potentially capture more nuanced word relationships.
* LSTM layer was used instead of SimpleRNN to help capture long-term dependencies in the data.
* Dropout layers were added after the embedding and LSTM layers to improve generalization by reducing overfitting.

The improvement in test accuracy from 0.795 in the original SimpleRNN model to 0.8278 in the modified LSTM model suggests that the changes enhanced the model's ability to capture important features in the IMDb dataset.

**Analysis of Changes and Observations:**

1. Embedding Size Increase: By raising the embedding dimension, the model could learn more detailed word representations. This improvement allows it to distinguish sentiments more effectively, especially in nuanced cases. The increased embedding dimension likely contributed to the accuracy boost, but it also adds computational cost.

2. LSTM Layer Replacement: Switching from SimpleRNN to LSTM has shown to be effective because LSTMs handle long-term dependencies better. For sentiment analysis, context and word dependencies over the review length are crucial, making LSTMs an ideal choice. This change likely provided a more meaningful improvement in classification accuracy.

3. Dropout Layers: Adding dropout layers helped reduce overfitting by introducing regularization. By preventing certain neurons from over-relying on specific features during training, dropout forces the network to generalize better. This likely contributed to the higher test accuracy.

These adjustments led to a stronger model capable of understanding the underlying sentiment in movie reviews with greater accuracy. The LSTM's ability to maintain information over longer sequences helped capture sentiment context more effectively than the SimpleRNN model.